src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java

```diff
@@ -126,41 +113,49 @@
     @Override
     public void beginTree(DetailAST rootAST) {
         publicTypeFound = false;
-        lineNumberTypeMap.clear();
 
         DetailAST currentNode = rootAST;
         while (currentNode != null) {
-            if (currentNode.getType() == TokenTypes.CLASS_DEF
-                    || currentNode.getType() == TokenTypes.ENUM_DEF
-                    || currentNode.getType() == TokenTypes.INTERFACE_DEF) {
-                if (isPublic(currentNode)) {
-                    publicTypeFound = true;
-                }
-                else {
-                    final String typeName = currentNode
-                            .findFirstToken(TokenTypes.IDENT).getText();
-                    lineNumberTypeMap.put(currentNode, typeName);
-                }
+            if (isTypeDef(currentNode) && isPublic(currentNode)) {
+                publicTypeFound = true;
+                break;
             }
             currentNode = currentNode.getNextSibling();
         }
     }
 
     @Override
     public void finishTree(DetailAST rootAST) {
-        if (!lineNumberTypeMap.isEmpty()) {
-            if (!publicTypeFound) {
-                // skip first top-level type.
-                lineNumberTypeMap.remove(lineNumberTypeMap.firstKey());
-            }
-
-            for (Map.Entry<DetailAST, String> entry
-                    : lineNumberTypeMap.entrySet()) {
-                log(entry.getKey(), MSG_KEY, entry.getValue());
+        DetailAST currentNode = rootAST;
+        boolean firstTypeSkipped = false;
+        while (currentNode != null) {
+            if (isTypeDef(currentNode) && !isPublic(currentNode)) {
+                if (publicTypeFound || firstTypeSkipped) {
```

Comment:
@romani 
> Checks that each top-level class, interface or enum resides in a source file of its own.

If we have 2 public classes in the same file, shouldn't we produce a violation on one of them since there is more than 1 top level class?

````
$ cat TestClass.java
public class TestClass1 {
    void method() {
    }
}
public class TestClass2 {
    void method() {
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="OneTopLevelClassCheck" />
    </module>
</module>

$ java -jar checkstyle-8.30-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

Shouldn't we expect a violation on line 5?

Reply:
\>1 public type will give a compile error, but I do agree that the current behavior is counterintuitive.

If you want, I can fix it here or open another issue.

Reply:
Sorry, I didn't realize it is not compilable.